src/ai/ai-bedrock-cohere-embedding.ts

```diff
@@ -0,0 +1,155 @@
+/* eslint-disable no-console */
+import {
+  BedrockRuntimeClient,
+  InvokeModelCommand,
+} from '@aws-sdk/client-bedrock-runtime';
+import { Tiktoken } from 'tiktoken/lite';
+import { TokenTextSplitter } from '@langchain/textsplitters';
+import cl100k from 'tiktoken/encoders/cl100k_base.json';
+import type {
+  CohereEmbeddingBody,
+  CohereEmbeddingResponse,
+} from './bedrock-embedding-service-schemas.js';
+
+if (!process.env.AWS_REGION) {
+  throw new Error('Please set AWS_REGION');
+}
+if (!process.env.AWS_ACCESS_KEY_ID) {
+  throw new Error('Please set AWS_ACCESS_KEY_ID');
+}
+if (!process.env.AWS_SECRET_ACCESS_KEY) {
+  throw new Error('Please set AWS_SECRET_ACCESS_KEY');
+}
+
+const bedrockClient = new BedrockRuntimeClient({
+  region: process.env.AWS_REGION,
+  credentials: {
+    accessKeyId: process.env.AWS_ACCESS_KEY_ID as string,
+    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY as string,
+  },
+});
+
+/**
+ * Verify that the string sent has less tokens than the maximum possible for the model
+ * @param root named parameters
+ * @param root.chunk the chunk to verify
+ * @returns true or false
+ */
+function exceedsTokenLimit({ chunk }: { chunk: string }): boolean {
+  const encoding = new Tiktoken(
+    cl100k.bpe_ranks,
+    cl100k.special_tokens,
+    cl100k.pat_str,
+  );
+  const tokens = encoding.encode(chunk);
+  encoding.free();
+  // cohere input token limit = 512
+  return tokens.length > 512;
+}
+
+/**
+ * Call the bedrock embedding model
+ * @param root named parameters
+ * @param root.chunks A document, or a list of chunks from a document to embed
+ * @param root.isQuery is the documnt for query
+ * @returns an object with the averaged embedding an an array of chunk embeddings
+ */
+export async function invokeBedrockCohereEmbedding({
+  chunks,
+  isQuery,
+}: {
+  chunks: string[];
+  isQuery: boolean;
+}): Promise<CohereEmbeddingResponse[] | undefined> {
+  const processedChunks = await Promise.all(
+    chunks.map(async (chunk) => {
+      return chunk.replace('\n', '');
+    }),
+  );
+  try {
+    if (
+      chunks.some((chunk) => {
+        return exceedsTokenLimit({ chunk });
+      })
+    ) {
+      throw new Error('A document exceeds token limit');
+    }
+    const embeddingResponses = Promise.all(
+      processedChunks.map(async (chunk) => {
```

Comment:
I'm worried this bit of the code could reach the rate limitation

Can you confirm:

- [ ] What is the rate limitation
- [ ] If the SDK help you avoid it (by limiting the number of requests, slowing them down)

If there is a rate limit and the SDK doesn't slow down your requests, you will need to slow them down yourself, AND remove the async map (because it will do it in parallel)